rfcs/0002-validation/text.md

```diff
@@ -0,0 +1,149 @@
+2. A dataset file in the staging bucket with a `x-amz-meta-dataset-id` metadata field whose value is the `id` of the new
+   row in the `dataset` table.
+
+A [CloudWatch Event](https://docs.aws.amazon.com/AmazonCloudWatch/latest/events/CloudWatch-Events-tutorial-ECS.html)
+will trigger the execution of the Fargate task when a new dataset file is added to the staging bucket.
+
+### Steps of the Fargate Task
+
+1. Check the `dataset` table to verify that the expected row exists and `upload_status` is "uploaded". To avoid possible
+   race conditions from the upload process, wait and recheck if this isn't true.
+2. Set the `validation_status` in `dataset` to "validating".
+3. Localize the uploaded file. HDF5 doesn't usually work with remote files very well.
+4. Run the `cellxgene schema validate` command to do a shallow validation of the dataset file.
+   - If this has a non-zero exit code, set `validation_status` to "invalid", write stdout from the command to
+     `processing_message`, and halt.
+   - Otherwise, set `validation_status` to "valid" and `artifact_status` to "creating".
+5. Read the appropriate values from the dataset file and upload the JSONB fields in the `dataset` table for tissue,
+   assay, etc.
+6. For format in `{loom, h5ad, rds}`
+   - If the format does not match the format of the uploaded dataset file, run a conversion script.
+   - Copy the file to the artifact bucket.
+   - Insert a row into `artifacts` with the format and S3 key.
+7. Convert to cxg, copy to the cellxgene deployment bucket, and insert a row into `deployment_directories`.
+8. Set `artifact_status` to "complete".
+
+The format conversions in steps 6 and 7 can occur in parallel, but they're also slower and sometimes use a lot of
+memory. There are existing tools that can perform the conversions: [sceasy](https://github.com/cellgeni/sceasy),
+[commands in
+Seurat](https://satijalab.org/seurat/v3.0/conversion_vignette.html), and `cellxgene convert`.
+
+The main step that impacts submitter user experience is 4. Users will want to know as quickly as possible whether their
+file has been accepted, so this is updated in the DB before any other work takes place.
+
+### Monitoring and error reporting
+
+The initial volume of submitted datasets is expected to be relatively low, on the order of tens per month. So we likely
+will not need a very complex monitoring solution at first.
+
+There are two main types of errors we should anticipate. The first is a failure of the submitted dataset to validate.
+This isn't quite an "error" as it's part of the expected functioning of the system, but it nevertheless results in a
+dataset that is not properly loaded into the portal. In that case, the dataset is marked as "invalid" in the database
+along with an error message that the frontend can display to the user.
+
+The other is an unexpected failure during the Fargate task execution. In the initial implementation, these can be found
```

Comment:
Can you add alerting so that we know when we need to manually restart or something?

Reply:
Yeah, though alert to where? Just ping the slack channel or something?

Reply:
If we add this to the upload step function we could automate the retry if it fails. Just a thought.